### ISSUE TITLE
Improve exception handling

### ISSUE THREAD

**Issue by AaronVanGeffen** [OpenLoco/OpenLoco#823]
Currently, OpenLoco primary exception handler is located in the main function. It is very basic:
https://github.com/OpenLoco/OpenLoco/blob/2963fd4475e92a24aa21f0c25a9395ef91ce026a/src/OpenLoco/OpenLoco.cpp#L1086-L1089

This is fine for devs working with a debugger console open. However, a regular user doesn't pay attention to the console. Therefore, due to this way of handling exceptions, they will likely never see the exception! Instead, all they see is the game closing unexpectedly. This needs to change.

My suggestion is to call a platform function in this location. This function would open a message box, showing a human-readable error message, along with the exception text for debugging purposes.

**Comment by TransshipmentEnvoy** [comment]
Will things like this be enough?
```
@@ -1182,6 +1182,7 @@ namespace OpenLoco
         catch (const std::exception& ex)
         {
             std::cerr << ex.what() << std::endl;
+            Ui::showMessageBox("caught exception", ex.what());
         }
     }
 }
```

**Comment by AaronVanGeffen** [comment]
It cannot be shown in-game anymore at this point; the game is quitting. Showing a message box will require a different implementation for Windows, macOS, and Linux. My thinking is that this is best delegated to some kind of platform function. You could start with just a Windows implementation, and stub out the macOS and Linux ones, though.

**Comment by TransshipmentEnvoy** [comment]
> Showing a message box will require a different implementation for Windows, macOS, and Linux.

I think ```Ui::showMessageBox``` calls ```SDL_ShowSimpleMessageBox```, it should be available on Windows, macOS and Linux? I do not quite understand why a platform specific message box implementation is required here for reporting exception to users.

**Comment by AaronVanGeffen** [comment]
> I think `Ui::showMessageBox` calls `SDL_ShowSimpleMessageBox`, it should be available on Windows, macOS and Linux? I do not quite understand why a platform specific message box implementation is required here for reporting exception to users.

You're right, that SDL function looks like it will do.

> Do you mean that a message box should be shown where the exceptions are thrown rather than when they are caught in `OpenLoco::main()`?

No, I meant we can't just paint a message box on the game canvas / UI. The exception is finally caught outside of the main loop, so the UI drawing functions will no longer be called.

**Comment by AaronVanGeffen** [comment]
Implemented by #1104
